File: frontend/identity-admin/src/data/identity-list-query.ts
export interface IdentityListQuery {
  page: number
  pageSize: number
  search?: string
  searchFields?: string[]
  filters?: Record<string, string>
  sort?: Array<{ field: string; direction: 'asc' | 'desc' }>
}

export interface IdentityPage<T> {
  items: T[]
  page: number
  page_size: number
  total: number
  has_next: boolean
}

export function identityListParams(query: IdentityListQuery): string {
  const params = new URLSearchParams({
    page: String(query.page),
    page_size: String(query.pageSize),
  })
  if (query.search?.trim()) params.set('search', query.search.trim())
  if (query.searchFields?.length) params.set('search_fields', query.searchFields.join(','))
  if (query.filters && Object.keys(query.filters).length) params.set('filters', JSON.stringify(query.filters))
  if (query.sort?.length) params.set('sort', query.sort.map((rule) => `${rule.field}:${rule.direction}`).join(','))
  return params.toString()
}

File: internal/domain/identity/model/identity_list_query.go
package identitymodel

// IdentitySortRule describes a stable ordering for Identity directory pages.
type IdentitySortRule struct {
	Field     string `json:"field"`
	Direction string `json:"direction"`
}

// IdentityListQuery is the query contract shared by Identity directory,
// workforce, role, and assignment searches. It is intentionally independent
// from host business-record filtering and transaction semantics.
type IdentityListQuery struct {
	AfterID                   string             `json:"after_id,omitempty"`
	PageSize                  int                `json:"page_size"`
	Search                    string             `json:"search,omitempty"`
	SearchFields              []string           `json:"search_fields,omitempty"`
	Filters                   map[string]any     `json:"filters,omitempty"`
	Sort                      []IdentitySortRule `json:"sort,omitempty"`
	Scope                     string             `json:"-"`
	PrincipalUserID           string             `json:"-"`
	PrincipalDepartmentPath   string             `json:"-"`
	PrincipalReportingUserIDs []string           `json:"-"`
}
